src/webgl/primitives.js

```diff
@@ -885,18 +885,20 @@ p5.RendererGL.prototype.curve=function
  * </code>
  * </div>
  */
-p5.RendererGL.prototype.line = function(x0,y0,z0,x1,y1,z1) {
-  if (typeof x0 !== 'undefined' ||
-      typeof y0 !== 'undefined' ||
-      typeof z0 !== 'undefined' ||
-      typeof x1 !== 'undefined' ||
-      typeof y1 !== 'undefined' ||
-      typeof z1 !== 'undefined')
-  {
+p5.RendererGL.prototype.line = function() {
+  if(arguments.length === 6) {
     this.beginShape();
-    this.vertex(x0, y0, z0);
-    this.vertex(x1, y1, z1);
+    this.vertex(arguments[0], arguments[1], arguments[2]);
+    this.vertex(arguments[3], arguments[4], arguments[5]);
     this.endShape();
+  } else if(arguments.length === 4) {
+    this.beginShape();
+    this.vertex(arguments[0], arguments[1], 0);
+    this.vertex(arguments[2], arguments[3], 0);
+    this.endShape();
+  } else {
```

Comment:
sorry I'm still not understanding why this last part is necessary. shouldn't the validateParameters check in the top level line() function catch the case where line is called without either 4 or 6 arguments?

Reply:
Hi @lmccart - no worries. I'll try do a better job of explaining myself in future!

There are actually two separate issues here.

First - the issue of the validateParameters() call. Why is there a need for a parameter check if validateParameters() is _already_ being called?

The problem is that the validateParameters check in the top level line _won't_ catch the cases where more than 4, but not 6, parameters are passed to line() in 3D.
*If* the first 4 arguments passed in are numbers, *then* the validation passes.
This is a known issue with the FES - it's the second point in [this section in the docs](https://github.com/processing/p5.js/wiki/Friendly-Error-System#known-limitations).

I'm guessing that this is what the original series of checks that I flagged were trying to address - they just got the logic wrong (|| instead of &&).

Second - there's this extra condition I've added.
The reason I've added this condition for a 4 parameter call is that I'm assuming that if someone calls line() with 4 parameters, they're expecting the behavior that they'd get if they were in the 2D drawing mode. Namely, that the four parameters represent two 2d points.
In the case of 4 parameters passed to the line() function in 3D, I'm just popping in a 0 for the z-coordinate.

So if we drop this "else if (arguments.length === 4)" condition I've added, then line() in 3D will only ever work with 6 parameters, which is 100%.
It's only there to support the usual 2D case while drawing in 3D - more than happy to remove it though.